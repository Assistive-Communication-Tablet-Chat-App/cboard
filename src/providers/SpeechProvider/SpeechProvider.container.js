import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import tts from './tts';
import { getVoices, changeVoice } from './SpeechProvider.actions';
import {
  changeLang,
  setLangs
} from '../LanguageProvider/LanguageProvider.actions';
import { DEFAULT_LANG } from '../../components/App/App.constants';
import { getVoicesLangs } from '../../i18n';

export class SpeechProvider extends Component {
  static propTypes = {
    langs: PropTypes.array.isRequired,
    children: PropTypes.node.isRequired
  };

  async componentDidMount() {
    const {
      lang: propsLang,
      setLangs,
      changeLang,
      voiceURI,
      changeVoice
    } = this.props;
    if (tts.isSupported()) {
      const voices = await this.props.getVoices();
      let supportedLangs = [DEFAULT_LANG];
      if (voices.length) {
        const sLanguages = getVoicesLangs(voices);
        if (sLanguages !== undefined && sLanguages.length) {
          supportedLangs = sLanguages;
        }
      }
      // hack just for alfanum voice
      if (
        supportedLangs.length &&
        (supportedLangs.includes('sr-RS') || supportedLangs.includes('sr-ME'))
      ) {
        supportedLangs.push('me-ME');
      }
      const lang = supportedLangs.includes(propsLang)
        ? propsLang
        : this.getDefaultLang(supportedLangs);
      setLangs(supportedLangs);
      changeLang(lang);

      const uris = voices.map(v => {
        return v.voiceURI;
      });
      if (uris.includes(voiceURI)) {
        changeVoice(voiceURI, lang);
      }
    }
  }

  getDefaultLang(langs) {
    for (let i = 0; i < langs.length; i++) {
      if (window.navigator.language.slice(0, 2) === langs[i].slice(0, 2)) {
        return langs[i];
      }
    }
    return langs.includes(DEFAULT_LANG) ? DEFAULT_LANG : langs[0];
  }

  render() {
    const { children } = this.props;

    return React.Children.only(children);
  }
}

const mapStateToProps = state => ({
  langs: state.speech.langs,
  lang: state.language.lang,
  voiceURI: state.speech.options.voiceURI
});

const mapDispatchToProps = {
  getVoices,
  changeLang,
  setLangs,
  changeVoice
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(SpeechProvider);
