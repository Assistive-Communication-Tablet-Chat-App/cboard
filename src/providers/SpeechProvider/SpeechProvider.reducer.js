import {
  RECEIVE_VOICES,
  CHANGE_VOICE,
  CHANGE_PITCH,
  CHANGE_RATE,
  CHANGE_VOLUME,
  START_SPEECH,
  END_SPEECH,
  EMPTY_VOICES
} from './SpeechProvider.constants';
import { getVoiceURI } from '../../i18n';
import { CHANGE_LANG } from '../LanguageProvider/LanguageProvider.constants';
import { LOGIN_SUCCESS } from '../../components/Account/Login/Login.constants';
import { DEFAULT_LANG } from '../../components/App/App.constants';

const initialState = {
  voices: [],
  langs: [],
  options: {
    lang: '',
    voiceURI: null,
    pitch: 1.0,
    rate: 1.0,
    volume: 1
  },
  isSpeaking: false
};

function speechProviderReducer(state = initialState, action) {
  switch (action.type) {
    case LOGIN_SUCCESS:
      const settings = action.payload.settings || {};
      const { speech } = settings || {};

      const pitch = speech && speech.pitch ? speech.pitch : state.options.pitch;
      const rate = speech && speech.rate ? speech.rate : state.options.rate;

      const options = { ...state.options, pitch, rate };

      return {
        ...state,
        options
      };
    case RECEIVE_VOICES:
      return {
        ...state,
        voices: action.voices,
        langs: [...new Set(action.voices.map(voice => voice.lang))].sort()
      };
    case CHANGE_VOICE:
      return {
        ...state,
        options: {
          ...state.options,
          voiceURI: action ? action.voiceURI : EMPTY_VOICES,
          lang: action ? action.lang : DEFAULT_LANG
        }
      };
    case CHANGE_LANG:
      // hack just for alfanum voice
      if (
        action.lang === 'sr-RS' ||
        action.lang === 'sr-ME' ||
        action.lang === 'me-ME'
      ) {
        return {
          ...state,
          options: {
            ...state.options,
            lang: action.lang,
            voiceURI: getVoiceURI(action.lang, state.voices)
          },
          langs: ['me-ME', 'sr-ME', 'sr-RS']
        };
      } else {
        return {
          ...state,
          options: {
            ...state.options,
            lang: action.lang,
            voiceURI: getVoiceURI(action.lang, state.voices)
          }
        };
      }
    case CHANGE_PITCH:
      return { ...state, options: { ...state.options, pitch: action.pitch } };
    case CHANGE_RATE:
      return { ...state, options: { ...state.options, rate: action.rate } };
    case CHANGE_VOLUME:
      return { ...state, options: { ...state.options, volume: action.volume } };
    case START_SPEECH:
      return { ...state, isSpeaking: action.isSpeaking };
    case END_SPEECH:
      return { ...state, isSpeaking: action.isSpeaking };
    default:
      return state;
  }
}

export default speechProviderReducer;
