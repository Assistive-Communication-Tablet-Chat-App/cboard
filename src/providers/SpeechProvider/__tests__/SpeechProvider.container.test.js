import { SpeechProvider } from '../SpeechProvider.container';
import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

jest.mock('react-redux');

describe('container group Test', () => {
  const props = {
    voices: [],
    langs: [],
    children: {}
  };
  it('speechProviderContainer: snapshot test', () => {
    const snapshot = renderer
      .create(
        <SpeechProvider {...props}>
          <div>child one</div>
        </SpeechProvider>
      )
      .toJSON();
    expect(snapshot).toMatchSnapshot();
  });
});
