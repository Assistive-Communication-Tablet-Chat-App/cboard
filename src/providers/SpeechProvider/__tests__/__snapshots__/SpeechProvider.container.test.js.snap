// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`container group Test speechProviderContainer: snapshot test 1`] = `
<div>
  child one
</div>
`;
